system/default.nix

```diff
@@ -21,48 +23,71 @@
     package = pkgs.nix;
   };
 
-  systemd.services.linger = {
-    enable = true;
-
-    requires = [ "local-fs.target" ];
-    after = [ "local-fs.target" ];
-
-    serviceConfig = {
-      Type = "oneshot";
-      ExecStart = ''
-        /run/current-system/sw/bin/loginctl enable-linger core
-      '';
-    };
-
-    wantedBy = [ "multi-user.target" ];
-  };
+  boot = {
+    kernelParams = [
+      "boot.shell_on_fail"
 
-  boot.kernelParams = [
-    "mitigations=off"
-    "l1tf=off"
-    "mds=off"
-    "no_stf_barrier"
-    "noibpb"
-    "noibrs"
-    "nopti"
-    "nospec_store_bypass_disable"
-    "nospectre_v1"
-    "nospectre_v2"
-    "tsx=on"
-    "tsx_async_abort=off"
-  ];
-  boot.kernelPackages = pkgs.linuxPackagesFor (pkgs.linux_latest);
+      "mitigations=off"
+      "l1tf=off"
+      "mds=off"
+      "no_stf_barrier"
+      "noibpb"
+      "noibrs"
+      "nopti"
+      "nospec_store_bypass_disable"
+      "nospectre_v1"
+      "nospectre_v2"
+      "tsx=on"
+      "tsx_async_abort=off"
+    ];
+    kernelPackages = lib.mkDefault (pkgs.linuxPackagesFor (pkgs.linux_latest));
+  } // (
+    # Increase tmpfs (default: "50%")
+    if (lib.trivial.release == "22.11") then {
+      tmpOnTmpfsSize = "80%";
+    } else {
+      tmp.tmpfsSize = "80%";
+    }
```

Comment:
Now that I think of it, does this work on Raspberry Pi as well? I am very surprised if so.

Reply:
The tmpfs _options_ (might use, but does not directly allocate; this is in fact the main feature of tmpfs over ramfs) a partition of the RAM for the post-boot filesystem. When the amount of RAM is low, say, 4GB, it would mean that only around 850MB is for any system resources, supposing that the tmpfs gets filled. You can fill the tmpfs by e.g. copying another initrd onto the filesystem, or run a nix build process. Now, the tmpfs may want to grow and evict system resources to a swap device. Often the system resources cannot be evicted without killing the processes, and might not fit into zram, so usually what happens instead is a system lock-down which makes the system unresponsive. When this happens, watchdog shuts down the server. This is related to https://github.com/ponkila/homestaking-infra/issues/9

Now that on the build servers we managed to overlay persistent Nix store onto the tmpfs, it might be that we can revert this change. That is to say, I don't know what is the best route of action here. I think we can keep this as is for now, but it might be a surprise to some users. I'm just bit surprised that seemingly this did not cause any problems on the RPi.